src/pybamm/solvers/processed_variable.py

```diff
@@ -76,45 +72,416 @@ def __init__(
         self.base_eval_size = self.base_variables[0].size
 
         # xr_data_array is initialized
-        self._xr_data_array = None
+        self._xr_array_raw = None
+        self._entries_raw = None
+        self._entries_for_interp_raw = None
+        self._coords_raw = None
+
+    def initialise(self):
+        if self.entries_raw_initialized:
+            entries = self._entries_raw
+            entries_for_interp = self._entries_for_interp_raw
+            coords = self._coords_raw
+        else:
+            entries = self.observe_raw()
+
+            t = self.t_pts
+            entries_for_interp, coords = self._interp_setup(entries, t)
+
+            self._entries_raw = entries
+            self._entries_for_interp_raw = entries_for_interp
+            self._coords_raw = coords
+
+    def observe_and_interp(self, t, fill_value):
+        """
+        Interpolate the variable at the given time points and y values.
+        t must be a sorted array of time points.
+        """
+
+        entries = self._observe_hermite_cpp(t)
+        processed_entries = self._observe_postfix(entries, t)
+
+        tf = self.t_pts[-1]
+        if t[-1] > tf and fill_value != "extrapolate":
+            # fill the rest
+            idx = np.searchsorted(t, tf, side="right")
+            processed_entries[..., idx:] = fill_value
+
+        return processed_entries
+
+    def observe_raw(self):
+        """
+        Evaluate the base variable at the given time points and y values.
+        """
+        t = self.t_pts
+
+        # For small number of points, use Python
+        if pybamm.has_idaklu() and (t.size > 1):
```

Comment:
1 is very small! So even for 10 or so points cpp is faster (I'm worried about the overhead of serialising and regenerating the casadi functions....)?

Reply:
C++ used to be slower for one entry, but now they are the same. Removing the size check.

To address the issue with serializing casadi functions, I added a pre-processing step in `_setup_cpp_inputs` to only serialize the subset of casadi functions that will actually be used